skbio/io/format/gff3.py

```diff
@@ -0,0 +1,534 @@
+def _dna_to_gff3(obj, fh, skip_subregion=True):
+    # write file header
+    fh.write('##gff-version 3\n')
+    _serialize_seq(obj, fh, skip_subregion)
+
+
+@gff3.reader(IntervalMetadata)
+def _gff3_to_interval_metadata(fh, seq_id):
+    '''Read a GFF3 record into the specified interval metadata.
+
+    Parameters
+    ----------
+    fh : file handler
+    '''
+    length = None
+    for data_type, sid, data in _yield_record(fh):
+        if seq_id == sid:
+            if data_type == 'length':
+                # get length from sequence-region pragma
+                length = data
+            elif data_type == 'data':
+                return _parse_record(data, length)
+            else:
+                raise GFF3FormatError(
+                    'Unknown section in the input GFF3 file: '
+                    '%r %r %r' % (data_type, sid, data))
+    # return an empty instead of None
+    return IntervalMetadata(None)
+
+
+@gff3.writer(IntervalMetadata)
+def _interval_metadata_to_gff3(obj, fh, seq_id, skip_subregion=True):
+    '''
+    Parameters
+    ----------
+    obj : IntervalMetadata
+    seq_id : str
+        ID for column 1 in the GFF3 file.
+    '''
+    # write file header
+    fh.write('##gff-version 3\n')
+    _serialize_interval_metadata(obj, seq_id, fh, skip_subregion=True)
+
+
+def _construct_seq(fh, constructor=DNA, seq_num=1):
+    lines = []
+    for i, (data_type, seq_id, l) in enumerate(_yield_record(fh), 1):
+        if data_type == 'data' and seq_num == i:
+            lines = l
+    # you can't read directly from fh because in registry.py line 543
+    # file.tell() will fail "telling position disabled by next() call".
+    stream = StringIO(fh.read())
+    seq = read(stream, format='fasta', into=constructor, seq_num=seq_num)
+    seq.interval_metadata = _parse_record(lines, len(seq))
+    return seq
+
+
+def _yield_record(fh):
+    '''Yield (seq_id, lines) that belong to the same sequence.'''
+    lines = []
+    current = False
+    for line in _line_generator(fh, skip_blanks=True, strip=True):
+        if line.startswith('##sequence-region'):
+            _, seq_id, start, end = line.split()
+            length = int(end) - int(start) + 1
+            yield 'length', seq_id, length
+        if line.startswith('##FASTA'):
+            # stop once reaching to sequence section
+            break
+        if not line.startswith('#'):
+            try:
+                seq_id, _ = line.split('\t', 1)
+            except ValueError:
+                raise GFF3FormatError(
+                    'Wrong GFF3 format at line: %s' % line)
+            if current == seq_id:
+                lines.append(line)
+            else:
+                if current is not False:
+                    yield 'data', current, lines
+                lines = [line]
+                current = seq_id
+    if current is False:
+        # if the input file object is empty, it should return
+        # an empty generator
+        return
```

Comment:
is the `return` necessary? somewhat surprised this is valid to have within a function that has `yield`